From e1222908e45c496b5553dc1c7353437d965d39fd Mon Sep 17 00:00:00 2001
From: Natanael Copa <ncopa@alpinelinux.org>
Date: Thu, 4 May 2023 13:12:35 +0200
Subject: [PATCH 4/9] nocloud: add meta-data network-interfaces and resolv_conf

Allow user to set a network-interfaces file in meta-data and set up dns
servers via resolv_conf.

This is for compatibility with cloud-init, and is a simple way to let
user configure static ip or advanced network

ref: https://cloudinit.readthedocs.io/en/latest/reference/datasources/nocloud.html#example-meta-data
ref: https://cloudinit.readthedocs.io/en/latest/reference/modules.html#resolv-conf
---
 lib/tiny-cloud/cloud/nocloud/init | 29 ++++++++++++++++++++++++++-
 tests/nocloud-init.test           | 33 +++++++++++++++++++++++++++++--
 2 files changed, 59 insertions(+), 3 deletions(-)

diff --git a/lib/tiny-cloud/cloud/nocloud/init b/lib/tiny-cloud/cloud/nocloud/init
index 56114e0..210d84e 100644
--- a/lib/tiny-cloud/cloud/nocloud/init
+++ b/lib/tiny-cloud/cloud/nocloud/init
@@ -76,6 +76,33 @@ set_network_config_auto() {
         "use dhcp"  >> "$ROOT"/etc/network/interfaces
 }
 
+add_once() {
+    local file="$1"
+    shift
+    for line; do
+        if ! grep -x -F "$line" "$file" 2>/dev/null; then
+            mkdir -p "${file%/*}"
+            printf "%s\n" "$line" >> "$file"
+        fi
+    done
+}
+
+set_resolv_conf() {
+    # resolv.conf
+    local nameservers="$(imds meta-data/resolv_conf/nameservers)"
+    for i in $nameservers; do
+        local server="$(imds meta-data/resolv_conf/nameservers/$i)"
+        add_once "$ROOT"/etc/resolv.conf "nameserver $server"
+    done
+}
+
 init__set_network_config() {
-    set_network_config_auto
+    local interfaces="$(imds meta-data/network-interfaces)"
+    mkdir -p "$ROOT"/etc/network
+    if [ -n "$interfaces" ]; then
+        printf "%s\n" "$interfaces" > "$ROOT"/etc/network/interfaces
+    elif ! [ -f "$ROOT"/etc/network/interfaces ]; then
+        set_network_config_auto
+    fi
+    set_resolv_conf
 }
diff --git a/tests/nocloud-init.test b/tests/nocloud-init.test
index a753ad1..4076553 100755
--- a/tests/nocloud-init.test
+++ b/tests/nocloud-init.test
@@ -4,13 +4,15 @@
 
 export PREFIX="$srcdir"
 export MOCK=echo
+export CLOUD=nocloud
 lib="$srcdir"/lib/tiny-cloud/cloud/nocloud/init
 
 init_tests \
 	ethernets \
 	find_first_interface_up \
 	auto_detect_ethernet_interface \
-	set_network_config_auto
+	set_network_config_auto \
+	set_network_config_network_interfaces
 
 ethernets_body() {
 	fake_interfaces lo br0 eth0 eth2 eth11
@@ -53,7 +55,7 @@ set_network_config_auto_body() {
 	echo up > sys/class/net/eth1/operstate
 
 	atf_check \
-		sh -c ". $lib; set_network_config_auto"
+		sh -c ". $lib; init__set_network_config"
 	atf_check \
 		-o match:"auto lo" \
 		-o match:"iface eth1" \
@@ -61,3 +63,30 @@ set_network_config_auto_body() {
 		cat etc/network/interfaces
 }
 
+set_network_config_network_interfaces_body() {
+	fake_metadata_nocloud <<-EOF
+		network-interfaces: |
+		  auto eth1
+		  iface eth1
+		    use dhcp
+
+		resolv_conf:
+		  nameservers:
+		    - 8.8.8.8
+		    - 8.8.4.4
+	EOF
+
+	atf_check \
+		sh -c ". $lib; init__set_network_config"
+	atf_check \
+		-o match:"auto eth1" \
+		-o match:"iface eth1" \
+		-o match:"use dhcp" \
+		cat etc/network/interfaces
+
+	atf_check \
+		-o match:"^nameserver 8.8.8.8$" \
+		-o match:"^nameserver 8.8.4.4$" \
+		cat etc/resolv.conf
+}
+
-- 
2.40.1

